name.abuchen.portfolio/src/name/abuchen/portfolio/datatransfer/pdf/TargobankPDFExtractor.java

```diff
@@ -58,413 +68,529 @@ public String getLabel()
+                        .filter(i -> AccountTransaction.Type.DIVIDENDS //
+                                        .equals(((AccountTransaction) i.getSubject()).getType()) || //
+                                        AccountTransaction.Type.TAXES //
+                                                        .equals(((AccountTransaction) i.getSubject()).getType())) //
                         .collect(Collectors.groupingBy(Item::getDate, Collectors.groupingBy(Item::getSecurity)));
 
-        // iterate transactions and combine data from dividend and tax document
-        // k is date and v is security
-        dividends.forEach((k, v) -> {
-            // key is security and transactions are the single transactions
-            v.forEach((key, transactions) -> {
-                if (transactions.size() == 1)
-                {
-                    // set correct date stored in attribute, if given in
-                    // security, leave "wrong" date otherwise
-                    AccountTransaction a1 = (AccountTransaction) transactions.get(0).getSubject();
-                    if (a1.getSecurity().getAttributes().get(new AttributeType(ATTRIBUTE_PAY_DATE)) != null)
-                    {
-                        a1.setDateTime((LocalDateTime) a1.getSecurity().getAttributes()
-                                        .get(new AttributeType(ATTRIBUTE_PAY_DATE)));
-                    }
-                }
-                else if (transactions.size() == 2)
+        sellTaxTransactions.forEach((k, v) -> {
+            v.forEach((security, transactions) -> {
+
+                // @formatter:off
+                // It is possible that several sell transactions exist on
+                // the same day without one or with several taxes transactions.
+                // 
+                // We simplify here only one sell transaction with one
+                // related taxes transaction.
+                // @formatter:on
+
+                if (transactions.size() == 2)
                 {
-                    // get both transactions and make the tax document a1
-                    // the tax document must have a tax unit
-                    AccountTransaction a1 = (AccountTransaction) transactions.get(0).getSubject();
-                    AccountTransaction a2;
+                    BuySellEntry sellTransaction = null;
+                    AccountTransaction taxTransaction = null;
 
-                    if (a1.getUnit(Unit.Type.TAX).isPresent())
+                    // Which transaction is the taxes and which the sell?
+                    try
                     {
-                        a2 = (AccountTransaction) transactions.get(1).getSubject();
+                        sellTransaction = (BuySellEntry) transactions.get(0).getSubject();
+                        taxTransaction = (AccountTransaction) transactions.get(1).getSubject();
                     }
-                    else
+                    catch (ClassCastException e1)
```

Comment:
It is not a good pattern to have the control flow via exceptions. Exceptions are expensive to create. And in this case, you could easily check with "instanceof" - you know there are only those two types available.